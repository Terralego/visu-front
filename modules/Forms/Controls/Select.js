import React from 'react';
import { MenuItem, Button } from '@blueprintjs/core';
import { Select } from '@blueprintjs/select';

export class WrapperSelect extends React.Component {
  constructor (props) {
    super(props);
    const { values } = this.props;
    this.state = {
      value: values[0],
    };
  }

  handleChange = value => {
    const { onChange } = this.props;
    this.setState({ value });
    onChange(value);
  }

  handleClick = value => () => this.handleChange(value);

  render () {
    const { label, values = [] } = this.props;
    const { value } = this.state;
    const { handleChange, handleClick } = this;

    return (
      <div>
        <p>{label}</p>
        <Select
          popoverProps={{ usePortal: false }}
          items={values}
          filterable={values.length > 9}
          itemRenderer={val => <MenuItem onClick={handleClick(val)} key={val} text={val} />}
          noResults={<MenuItem disabled text="No results." />}
          onItemSelect={handleChange}
        >
          <Button text={value} rightIcon="double-caret-vertical" />
        </Select>
      </div>
    );
  }
}

export default WrapperSelect;
