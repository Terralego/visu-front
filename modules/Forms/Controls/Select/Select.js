import React from 'react';
import PropTypes from 'prop-types';
import classnames from 'classnames';
import { MenuItem, Button, Position } from '@blueprintjs/core';
import { Select as BPSelect } from '@blueprintjs/select';
import uuid from 'uuid/v4';

import { preventEnterKeyPress } from '../../../../utils/event';
import formatValues from '../formatValues';

export class Select extends React.Component {
  static propTypes = {
    locales: PropTypes.shape({ noResults: PropTypes.string }),
    label: PropTypes.string,
    onChange: PropTypes.func,
    values: PropTypes.arrayOf(
      PropTypes.oneOfType([
        PropTypes.string,
        PropTypes.shape({
          label: PropTypes.string.isRequired,
          value: PropTypes.any.isRequired,
        }),
      ]),
    ).isRequired,
    placeholder: PropTypes.string,
    isSubmissionPrevented: PropTypes.bool,
  };

  static defaultProps = {
    locales: { noResults: 'No results.' },
    label: '',
    placeholder: 'Filterâ€¦',
    isSubmissionPrevented: true,
    onChange () {},
  };

  static getDerivedStateFromProps ({ values }) {
    if (!values) return null;

    return {
      values: formatValues(values),
    };
  }

  state = {
    values: [],
    query: '',
  };

  uuid = `select-${uuid()}`;

  handleChange = ({ value }) => {
    const { onChange } = this.props;
    onChange(value);
  };

  handleQueryChange = query => {
    this.setState({ query });
  };

  render () {
    const {
      label,
      locales: { noResults, emptySelectItem },
      placeholder,
      isSubmissionPrevented,
      value,
      className,
      inline,
      fullWidth,
      ...props
    } = this.props;

    const { values, query } = this.state;
    const { handleChange } = this;
    const filteredItems = query === ''
      ? values
      : values.filter(({ label: itemLabel }) =>
        itemLabel.toLowerCase().includes(query.toLowerCase()));

    return (
      <div
        className={classnames(
          'control-container',
          'control-container--select',
          { 'control-container--full-width': fullWidth },
        )}
        onKeyPress={isSubmissionPrevented ? preventEnterKeyPress : null}
        role="presentation"
      >
        <label
          htmlFor={this.uuid}
          className={classnames(
            'control-label',
            { 'control-label--inline': inline },
          )}
        >
          {label}
        </label>
        <BPSelect
          className={classnames('tf-select', className)}
          popoverProps={{ usePortal: false, position: Position.BOTTOM_LEFT, minimal: true }}
          items={filteredItems}
          filterable={values.length > 9}
          inputProps={{ placeholder }}
          onQueryChange={this.handleQueryChange}
          itemRenderer={({
            label: itemLabel, value: itemValue,
          }, {
            handleClick, modifiers: { matchesPredicate, ...modifiers },
          }) => (
            <div
              key={`${itemValue}${itemLabel}`}
              className={classnames({
                'control-container__item': true,
                'control-container__item--empty': itemValue === '',
              })}
            >
              <MenuItem
                {...modifiers}
                onClick={handleClick}
                text={itemLabel}
              />
            </div>
          )}
          noResults={<MenuItem disabled text={noResults} />}
          onItemSelect={handleChange}
          {...props}
        >
          <Button id={this.uuid} text={value || emptySelectItem} rightIcon="double-caret-vertical" />
        </BPSelect>
      </div>
    );
  }
}

export default Select;
