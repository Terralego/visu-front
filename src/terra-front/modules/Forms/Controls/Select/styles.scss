.control-container--select {

  &.control-container--full-width {
    .bp3-popover-target, div button {
      width: 100%;
      outline: none;
      margin: auto;
    }

    button {
      display: flex;
      justify-content: space-between;
    }
  }
  .control-label {
    display: block;
    line-height: 2rem;

    &--inline {
      display: inline;
      margin-right: 1rem;
    }
  }
}
