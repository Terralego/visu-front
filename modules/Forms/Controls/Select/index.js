import Select from './Select';
import './styles.scss';

export default Select;
