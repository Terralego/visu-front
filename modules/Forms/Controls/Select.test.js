import React from 'react';
import renderer from 'react-test-renderer';
import { mount, shallow } from 'enzyme';

import Select from './Select';

jest.mock('@blueprintjs/select', () => ({
  Select: function BPSelect ({ items }) {
    return (
      <select>
        {items.map(({ label, value }) => (
          <option key={`${value}${label}`} value={value}>{label}</option>
        ))}
      </select>
    );
  },
}));
jest.mock('@blueprintjs/core', () => ({
  Button () { return null; },
  MenuItem: function BPMenuItem () { return null; },
}));

it('should render correctly', () => {
  const tree = renderer.create((
    <Select
      label="Pwout"
      onChange={() => null}
      values={['foo']}
    />
  ));
  expect(tree.toJSON()).toMatchSnapshot();
});

it('should render correctly with query', () => {
  const tree = renderer.create((
    <Select
      label="Pwout"
      onChange={() => null}
      values={['foo', 'bar', 'foofoo', 'fou', 'fooooo']}
    />
  ));
  tree.getInstance().handleQueryChange('fo');
  expect(tree.toJSON()).toMatchSnapshot();
});

it('should update items', () => {
  const instance = new Select({ values: [], locales: { emptySelectItem: 'empty' } });

  instance.setState = jest.fn();
  const prevValues = instance.props.values;
  instance.props.values = ['foo'];
  instance.componentDidUpdate({ values: prevValues });
  expect(instance.setState).toHaveBeenCalledWith({
    items: [{
      value: '',
      label: 'empty',
    }, {
      value: 'foo',
      label: 'foo',
    }],
  });
});

it('should handle change', () => {
  const onChange = jest.fn();
  const instance = new Select({ onChange });
  instance.handleChange({ value: 'foo' });
  expect(onChange).toHaveBeenCalledWith('foo');
});

it('should render Item', () => {
  const wrapper = mount(
    <Select
      values={['foo']}
    />,
  );
  const ItemRenderer = ({ item, ...props }) =>
    wrapper.find('BPSelect').props().itemRenderer(
      item,
      {
        ...props,
        modifiers: { ...props.modifiers || {} },
      },
    );

  expect(shallow(
    <ItemRenderer
      item={{
        label: 'Foo',
        value: 'foo',
      }}
    />,
  ).props().className).toBe('control-container__item');
  expect(shallow(
    <ItemRenderer
      item={{
        label: 'empty',
        value: '',
      }}
    />,
  ).props().className).toBe('control-container__item control-container__item--empty');
});

it('should handle query change', () => {
  const instance = new Select();
  instance.setState = jest.fn();
  instance.handleQueryChange('foo');
  expect(instance.setState).toHaveBeenCalledWith({ query: 'foo' });
});
