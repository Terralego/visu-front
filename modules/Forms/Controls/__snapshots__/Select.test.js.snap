// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should render correctly 1`] = `
<div
  className="control-container"
  onKeyPress={[Function]}
  role="presentation"
>
  <p
    className="control-label"
  >
    Pwout
  </p>
  <select>
    <option
      value={null}
    >
      
    </option>
    <option
      value="foo"
    >
      foo
    </option>
  </select>
</div>
`;

exports[`should render correctly with query 1`] = `
<div
  className="control-container"
  onKeyPress={[Function]}
  role="presentation"
>
  <p
    className="control-label"
  >
    Pwout
  </p>
  <select>
    <option
      value="foo"
    >
      foo
    </option>
    <option
      value="foofoo"
    >
      foofoo
    </option>
    <option
      value="fou"
    >
      fou
    </option>
    <option
      value="fooooo"
    >
      fooooo
    </option>
  </select>
</div>
`;
